File: bindings/swift/Sources/PQCDyber/Signature.swift
// libpqc-dyber - Post-Quantum Cryptography Library
// Copyright (c) 2024-2026 Dyber, Inc.
// SPDX-License-Identifier: Apache-2.0 OR MIT

import Foundation
import Cpqc

/// Digital signature operations.
public final class Signature {

    /// Signature keypair result.
    public struct KeyPair {
        public let publicKey: Data
        public let secretKey: Data
    }

    private let handle: OpaquePointer

    /// Create a Signature context for the specified algorithm.
    /// - Parameter algorithm: Algorithm name, e.g. "ML-DSA-65".
    /// - Throws: `PQCError.unsupportedAlgorithm` if the algorithm is not available.
    public init(algorithm: String) throws {
        guard let h = pqc_sig_new(algorithm) else {
            throw PQCError.unsupportedAlgorithm(algorithm)
        }
        self.handle = h
    }

    deinit {
        pqc_sig_free(handle)
    }

    /// The algorithm name.
    public var algorithm: String {
        String(cString: pqc_sig_algorithm(handle))
    }

    /// Public key size in bytes.
    public var publicKeySize: Int {
        pqc_sig_public_key_size(handle)
    }

    /// Secret key size in bytes.
    public var secretKeySize: Int {
        pqc_sig_secret_key_size(handle)
    }

    /// Maximum signature size in bytes.
    public var maxSignatureSize: Int {
        pqc_sig_max_signature_size(handle)
    }

    /// NIST security level (1-5).
    public var securityLevel: Int {
        Int(pqc_sig_security_level(handle))
    }

    /// Whether this is a stateful signature scheme.
    public var isStateful: Bool {
        pqc_sig_is_stateful(handle) != 0
    }

    /// Generate a new keypair.
    /// - Returns: A `KeyPair` containing the public and secret keys.
    /// - Throws: `PQCError.operationFailed` on failure.
    public func keygen() throws -> KeyPair {
        var pk = Data(count: publicKeySize)
        var sk = Data(count: secretKeySize)

        let rc = pk.withUnsafeMutableBytes { pkPtr in
            sk.withUnsafeMutableBytes { skPtr in
                pqc_sig_keygen(handle,
                               pkPtr.bindMemory(to: UInt8.self).baseAddress!,
                               skPtr.bindMemory(to: UInt8.self).baseAddress!)
            }
        }
        guard rc == 0 else { throw PQCError.operationFailed(rc) }
        return KeyPair(publicKey: pk, secretKey: sk)
    }

    /// Sign a message.
    /// - Parameters:
    ///   - message: The message to sign.
    ///   - secretKey: The signer's secret key.
    /// - Returns: The signature as `Data`.
    /// - Throws: `PQCError.operationFailed` on failure.
    public func sign(message: Data, secretKey: Data) throws -> Data {
        var sigBuf = Data(count: maxSignatureSize)
        var sigLen = maxSignatureSize

        let rc = sigBuf.withUnsafeMutableBytes { sigPtr in
            message.withUnsafeBytes { msgPtr in
                secretKey.withUnsafeBytes { skPtr in
                    pqc_sig_sign(handle,
                                 sigPtr.bindMemory(to: UInt8.self).baseAddress!,
                                 &sigLen,
                                 msgPtr.bindMemory(to: UInt8.self).baseAddress!,
                                 message.count,
                                 skPtr.bindMemory(to: UInt8.self).baseAddress!)
                }
            }
        }
        guard rc == 0 else { throw PQCError.operationFailed(rc) }
        return sigBuf.prefix(sigLen)
    }

    /// Verify a signature.
    /// - Parameters:
    ///   - message: The message that was signed.
    ///   - signature: The signature to verify.
    ///   - publicKey: The signer's public key.
    /// - Returns: `true` if the signature is valid.
    public func verify(message: Data, signature: Data, publicKey: Data) -> Bool {
        let rc = message.withUnsafeBytes { msgPtr in
            signature.withUnsafeBytes { sigPtr in
                publicKey.withUnsafeBytes { pkPtr in
                    pqc_sig_verify(handle,
                                   msgPtr.bindMemory(to: UInt8.self).baseAddress!,
                                   message.count,
                                   sigPtr.bindMemory(to: UInt8.self).baseAddress!,
                                   signature.count,
                                   pkPtr.bindMemory(to: UInt8.self).baseAddress!)
                }
            }
        }
        return rc == 0
    }
}

File: bindings/ruby/lib/pqc_dyber/signature.rb
# libpqc-dyber - Post-Quantum Cryptography Library
# Copyright (c) 2024-2026 Dyber, Inc.
# SPDX-License-Identifier: Apache-2.0 OR MIT
#
# Signature class.

module PqcDyber
  class Signature
    attr_reader :algorithm

    # Create a Signature instance for the given algorithm.
    #
    # @param algorithm [String] e.g. "ML-DSA-65"
    # @raise [RuntimeError] if algorithm is not supported
    def initialize(algorithm)
      @algorithm = algorithm
      @handle = PqcDyber.pqc_sig_new(algorithm)
      raise "Unsupported signature algorithm: #{algorithm}" if @handle.null?

      ObjectSpace.define_finalizer(self, self.class._release(@handle))
    end

    def public_key_size
      PqcDyber.pqc_sig_public_key_size(@handle)
    end

    def secret_key_size
      PqcDyber.pqc_sig_secret_key_size(@handle)
    end

    def max_signature_size
      PqcDyber.pqc_sig_max_signature_size(@handle)
    end

    def security_level
      PqcDyber.pqc_sig_security_level(@handle)
    end

    def stateful?
      PqcDyber.pqc_sig_is_stateful(@handle) != 0
    end

    # Generate a keypair.
    #
    # @return [Hash] { public_key: String, secret_key: String }
    def keygen
      pk = FFI::MemoryPointer.new(:uint8, public_key_size)
      sk = FFI::MemoryPointer.new(:uint8, secret_key_size)
      rc = PqcDyber.pqc_sig_keygen(@handle, pk, sk)
      raise "Signature keygen failed: #{PqcDyber.pqc_status_string(rc)}" unless rc == 0

      { public_key: pk.read_bytes(public_key_size),
        secret_key: sk.read_bytes(secret_key_size) }
    end

    # Sign a message.
    #
    # @param message [String]
    # @param secret_key [String]
    # @return [String] the signature
    def sign(message, secret_key)
      sig_buf = FFI::MemoryPointer.new(:uint8, max_signature_size)
      sig_len = FFI::MemoryPointer.new(:size_t)
      sig_len.write(:size_t, max_signature_size)

      msg_buf = FFI::MemoryPointer.from_string(message)
      sk_buf = FFI::MemoryPointer.from_string(secret_key)

      rc = PqcDyber.pqc_sig_sign(@handle, sig_buf, sig_len, msg_buf, message.bytesize, sk_buf)
      raise "Signature sign failed: #{PqcDyber.pqc_status_string(rc)}" unless rc == 0

      actual_len = sig_len.read(:size_t)
      sig_buf.read_bytes(actual_len)
    end

    # Verify a signature.
    #
    # @param message [String]
    # @param signature [String]
    # @param public_key [String]
    # @return [Boolean] true if valid
    def verify(message, signature, public_key)
      msg_buf = FFI::MemoryPointer.from_string(message)
      sig_buf = FFI::MemoryPointer.from_string(signature)
      pk_buf = FFI::MemoryPointer.from_string(public_key)

      rc = PqcDyber.pqc_sig_verify(@handle, msg_buf, message.bytesize,
                                     sig_buf, signature.bytesize, pk_buf)
      rc == 0
    end

    # @private
    def self._release(handle)
      proc { PqcDyber.pqc_sig_free(handle) }
    end
  end
end
